misc/pragmas.md

```diff
@@ -0,0 +1,574 @@
+`rule` element of the grammar.
+
+There are no syntactic distinctions among these cases: if the rule
+shown is the first rule in the grammar, `my:color` might be supplying
+a default property value for `a`, or annotating the rule, or
+annotating the grammar as a whole. The difference must be carried by
+the semantics of the pragma: those defining the `my:color` pragma
+should be clear about what it means and what its scope is.
+
+### The brackets-QName proposal 
+
+In this proposal, pragmas take the form of a left square bracket, an
+optional mark, a QName, the pragma's data, and a right square bracket.
+Nested pairs of square brackets are allowed, so pragmas can nest
+arbitrarily deep.
+
+Pragmas can occur immediately before a terminal or nonterminal symbol,
+before the mark if any, or in the whitespace immediately before the
+full stop of a rule.
+
+In the XML form, the serialization of the pragma depends on its
+position and the mark following the opening bracket.
+
+* Pragmas occurring before the full stop of a rule are serialized as
+extension elements following the final `alt` of the right-hand side of
+the rule.
+
+* Pragmas occurring before a symbol are serialized as attributes if 
+marked `@` (ignoring marks on any nested pragmas). 
+
+* Pragmas occurring before a symbol are not serialized at all if
+marked `-`.
+
+* Pragmas occurring before a symbol are serialized as extension
+elements if unmarked or marked `^`.
+
+  The extension elements appear as the first children of the XML
+representation of the symbol. In the canonical XML representation, the
+pragma data is serialized as a 'pragma-data' element within the
+extension element, which will contain the pragma-data as a character
+sequence. Depending on the semantics of the pragma itself, the
+extension element may also contain additional elements. (See the
+examples in the final section of this paper.)
+
+For example:
+
+````
+    [my:color blue] a = b, [@my:flavor vanilla] c? [my:spin ...].
+````
+
+The corresponding XML form is:
+
+````
+    <rule name="a">
+        <my:color>blue</my:color>
+        <alt>
+            <nonterminal name="b"/>
+            <option>
+                <nonterminal name="c" my:flavor="vanilla"/>
+            </option>
+        </alt>
+        <my:spin>
+            <pragma-data>...</pragma-data>
+        </my:spin>
+    </rule>
+````
+	
+Annotations pertaining to a particular occurrence of a symbol appear 
+immediately before that occurrence, as here the `my:flavor` pragma. 
+
+Annotations which provide a nonterminal with default value for some
+property appear immediately before the nonterminal on the left-hand
+side of a rule, as here the `my:color` pragma.
+
+Annotations pertaining to a rule as a whole may occur either before
+the left-hand side of the rule, which may be preferred for
+light-weight annotations, or before the full stop at the end of the
+rule, which is preferable for long complicated annotations.
```

Comment:
Is this merely for convenience and neatness?

Is it true to say that there is no semantic difference between a full stop annotation and a rule annotation?  What would we lose if we just chose one?